From bf49ffd97ef0ef8a008d616490419dcd097c74a4 Mon Sep 17 00:00:00 2001
From: Daniel Baluta <daniel.baluta@nxp.com>
Date: Wed, 23 May 2018 11:29:01 +0300
Subject: [PATCH 4373/5242] MLK-17481-2: ARM64: dts: imx8qm: Enable DSP

commit  f6f38d16ff0f924311a536b257a5069d017af314 from
https://source.codeaurora.org/external/imx/linux-imx.git

i.mx8QM B0 comes with a DSP (placed in the VPU unit).

From the ARM core side DSP local memory (Inst/Data) is mapped at
0x55000000-0x55FFFFFF range.

DSP also uses code located in SDRAM mapping starting at 0x92400000.

While at it, move rpmsg_node up in order to have all reserved
areas sorted by address.

Signed-off-by: Daniel Baluta <daniel.baluta@nxp.com>
Reviewed-by: Shengjiu Wang <shengjiu.wang@nxp.com>
(cherry picked from commit 1eb4b2ee64c6f1fd7a5d6ceb1f019e876dbdfeb9)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 .../boot/dts/freescale/fsl-imx8qm-device.dtsi      |   41 ++++++++++++++++++++
 .../boot/dts/freescale/fsl-imx8qm-mek-domu.dts     |    1 +
 arch/arm64/boot/dts/freescale/fsl-imx8qm.dtsi      |   22 +++++++++--
 3 files changed, 60 insertions(+), 4 deletions(-)

diff --git a/arch/arm64/boot/dts/freescale/fsl-imx8qm-device.dtsi b/arch/arm64/boot/dts/freescale/fsl-imx8qm-device.dtsi
index 64e703b..b1066c8 100644
--- a/arch/arm64/boot/dts/freescale/fsl-imx8qm-device.dtsi
+++ b/arch/arm64/boot/dts/freescale/fsl-imx8qm-device.dtsi
@@ -902,6 +902,35 @@
 					};
 				};
 			};
+
+			pd_dsp_mu_A: PD_DSP_MU_A {
+				reg = <SC_R_MU_13A>;
+				#power-domain-cells = <0>;
+				power-domains =<&pd_audio>;
+				#address-cells = <1>;
+				#size-cells = <0>;
+
+				pd_dsp_mu_B: PD_DSP_MU_B {
+					reg = <SC_R_MU_13B>;
+					#power-domain-cells = <0>;
+					power-domains =<&pd_dsp_mu_A>;
+					#address-cells = <1>;
+					#size-cells = <0>;
+
+					pd_dsp_ram: PD_AUD_OCRAM {
+						reg = <SC_R_DSP_RAM>;
+						#power-domain-cells = <0>;
+						power-domains =<&pd_dsp_mu_B>;
+						#address-cells = <1>;
+						#size-cells = <0>;
+						pd_dsp: PD_AUD_DSP {
+							reg = <SC_R_DSP>;
+							#power-domain-cells = <0>;
+							power-domains =<&pd_dsp_ram>;
+						};
+					};
+				};
+			};
 		};
 
 		pd_dma: PD_DMA {
@@ -3606,6 +3635,18 @@
 		status = "disabled";
 	};
 
+	dsp: dsp@556e8000 {
+		compatible = "fsl,imx8qxp-dsp";
+		reserved-region = <&dsp_reserved>;
+		reg = <0x0 0x556e8000 0x0 0x88000>;
+		clocks = <&clk IMX8QM_AUD_DSP_IPG>,
+			<&clk IMX8QM_AUD_OCRAM_IPG>,
+			<&clk IMX8QM_AUD_DSP_CORE_CLK>;
+		clock-names = "ipg", "ocram", "core";
+		fsl,dsp-firmware = "imx/dsp/hifi4.bin";
+		power-domains = <&pd_dsp>;
+	};
+
 	esai0: esai@59010000 {
 		compatible = "fsl,imx8qm-esai";
 		reg = <0x0 0x59010000 0x0 0x10000>;
diff --git a/arch/arm64/boot/dts/freescale/fsl-imx8qm-mek-domu.dts b/arch/arm64/boot/dts/freescale/fsl-imx8qm-mek-domu.dts
index d83c37c..973f875 100644
--- a/arch/arm64/boot/dts/freescale/fsl-imx8qm-mek-domu.dts
+++ b/arch/arm64/boot/dts/freescale/fsl-imx8qm-mek-domu.dts
@@ -331,6 +331,7 @@
 /delete-node/ &dpr2_channel2;
 /delete-node/ &dpr2_channel3;
 /delete-node/ &dpu1;
+/delete-node/ &dsp;
 /delete-node/ &irqsteer_dsi1;
 /delete-node/ &i2c0_mipi_dsi1;
 /delete-node/ &mipi_dsi_csr2;
diff --git a/arch/arm64/boot/dts/freescale/fsl-imx8qm.dtsi b/arch/arm64/boot/dts/freescale/fsl-imx8qm.dtsi
index ccdebea..ce6e431 100644
--- a/arch/arm64/boot/dts/freescale/fsl-imx8qm.dtsi
+++ b/arch/arm64/boot/dts/freescale/fsl-imx8qm.dtsi
@@ -81,6 +81,16 @@
 		 *
 		 */
 
+		rpmsg_reserved: rpmsg@0x90000000 {
+			no-map;
+			reg = <0 0x90000000 0 0x400000>;
+		};
+
+		dsp_reserved: dsp@0x92400000 {
+			no-map;
+			reg = <0 0x92400000 0 0x2000000>;
+		};
+
 		/* global autoconfigured region for contiguous allocations */
 		linux,cma {
 			compatible = "shared-dma-pool";
@@ -90,10 +100,6 @@
 			linux,cma-default;
 		};
 
-		rpmsg_reserved: rpmsg@0x90000000 {
-			no-map;
-			reg = <0 0x90000000 0 0x400000>;
-		};
 	};
 
 	gic: interrupt-controller@51a00000 {
@@ -119,6 +125,14 @@
 		status = "okay";
 	};
 
+	mu13: mu13@5d280000 {
+		compatible = "fsl,imx8-mu-dsp";
+		reg = <0x0 0x5d280000 0x0 0x10000>;
+		interrupts = <GIC_SPI 192 IRQ_TYPE_LEVEL_HIGH>;
+		fsl,dsp_ap_mu_id = <13>;
+		status = "okay";
+	};
+
 	clk: clk {
 		compatible = "fsl,imx8qm-clk";
 		#clock-cells = <1>;
-- 
1.7.9.5

